### DISCUSSION TITLE
Declaring supported TypeScript versions

### DISCUSSION THREAD

**Discussion by SimonAlling** [typed-ember/semver#23] [Ideas]
It might be nice if it were possible for a package to declare the minimum TypeScript version that its consumers must use.

The closest I've come to that is to use `typesVersions` to conditionally serve type declarations that deliberately don't compile under TypeScript versions other than those I support. For example:

## `minimum-typescript-version.ts`

```typescript
/**
 * @module
 * This module is intended to enforce the minimum TypeScript version we support, by not compiling under earlier versions.
 */

/**
 * Const type parameters were introduced in TypeScript 5.0.
 *
 * Under TypeScript 4.9:
 *
 * > error TS1139: Type parameter declaration expected.
 */
export default function<const _>(_: "TypeScript 5.0 or later is required.") {}
```

## `package.json`

```json
{
  "name": "the-package",
  "typesVersions": {
    "<5.0": {
      "*": [
        "./dist/minimum-typescript-version.d.ts"
      ]
    },
    "*": {
      …
    }
  },
}
```

A TypeScript 4.9 consumer that attempts to import _any_ export from the package in question, whether that export contains TypeScript 5.0-only syntax or not, will get this error:

<details>

<summary>Complete error message</summary>

```
node_modules/the-package/dist/minimum-typescript-version.d.ts:12:26 - error TS1139: Type parameter declaration expected.

12 export default function <const _>(_: "TypeScript 5.0 or later is required."): void;
                            ~~~~~

node_modules/the-package/dist/minimum-typescript-version.d.ts:12:33 - error TS1005: ',' expected.

12 export default function <const _>(_: "TypeScript 5.0 or later is required."): void;
                                   ~

node_modules/the-package/dist/minimum-typescript-version.d.ts:12:36 - error TS1005: ')' expected.

12 export default function <const _>(_: "TypeScript 5.0 or later is required."): void;
                                      ~

node_modules/the-package/dist/minimum-typescript-version.d.ts:12:76 - error TS1005: ';' expected.

12 export default function <const _>(_: "TypeScript 5.0 or later is required."): void;
                                                                              ~

node_modules/the-package/dist/minimum-typescript-version.d.ts:12:77 - error TS1128: Declaration or statement expected.

12 export default function <const _>(_: "TypeScript 5.0 or later is required."): void;
                                                                               ~

node_modules/the-package/dist/minimum-typescript-version.d.ts:12:83 - error TS1109: Expression expected.

12 export default function <const _>(_: "TypeScript 5.0 or later is required."): void;
                                                                                     ~
```

</details>

That means one can make it obvious to consumers if they have a non-supported TypeScript version, instead of flooding them with the wall of text that TypeScript tends to print in the presence of syntax errors.

This scenario is particularly interesting:

  1. `the-package` drops support for TypeScript <5.0 in version 42.0.0, but doesn't yet use `const` type assertions, so the emitted code still happens to work under older TypeScript versions.
  2. A consumer using TypeScript 4.9 upgrades `the-package` to version 42.0.0 and notices no breakage, so they think everything is fine.
  3. `the-package` starts using `const` type assertions in version 42.1.0. Now the emitted code suddenly does need TypeScript 5.0 or later to work.
  4. The consumer upgrades `the-package` to version 42.1.0 (maybe indirectly via a transitive dependency) and suddenly gets curious syntax errors.

To prevent that, it might be nice if version 42.0.0 really _did_ break consumers' builds, in a clear and descriptive way.

On the other hand, the above approach is quite hacky, and it may mean that some consumers won't be able to use the package, even though it — or at least some of its exports — would have worked perfectly fine.

But I think it's worth having a discussion around this at least: is it possible to declare and/or enforce a minimum supported TypeScript version, and should packages do that?

I might just be overthinking this of course. Hopefully someone will enlighten me!

**Comment by chriskrycho** [comment]
Just had time to come back to this – I think supporting this programmatically is a good idea (along with the “here are the supported versions” in the README etc.). One reasonably good option here is to make it an optional peer dependency (though support for that is relatively limited). Another is to use (possibly with some extension/further development) something like @rwjblue's [validate-peer-dependencies](https://github.com/rwjblue/validate-peer-dependencies) package. Notionally, the idea would be to make it easy for *tooling* to detect the issue and report it in a user-actionable way.

As an aside, this is part of why the spec says (or at least implies? I need to check!) that *depending* on a new feature is actually breaking unless you do some kind of down-leveling to provide support to earlier versions. In some cases, e.g. use of the module import spec in `package.json`, it may not even be possible to correctly support earlier versions—so adopting that would by definition be a breaking change.
